File: JKNoticationHelper/Classes/JKFastNotificationProxy.m
//
//  JKFastNotificationProxy.m
//  lib
//
//  Created by Demi on 7/10/16.
//  Copyright © 2016 Taylor. All rights reserved.
//

#import "JKFastNotificationProxy.h"

@interface JKFastNotificationProxy ()

@property (nonatomic, copy) NSString *notificationName;
@property (nonatomic, copy) JKFastNotificationBlock block;

@end

@implementation JKFastNotificationProxy

- (instancetype)initWithName:(NSString *)name block:(JKFastNotificationBlock)block
{
    self = [super init];
    if (self)
    {
        self.notificationName = name;
        self.block = block;
        [[NSNotificationCenter defaultCenter] addObserver:self
                                                 selector:@selector(receivedNotifcation:)
                                                     name:_notificationName
                                                   object:nil];
    }
    return self;
}

- (void)receivedNotifcation:(NSNotification *)note
{
    if (_block)
    {
        _block(note);
    }
}

- (void)dealloc
{
    [[NSNotificationCenter defaultCenter] removeObserver:self name:_notificationName object:nil];
}

@end

File: Swift/JKNoticationHelper_Swift/Classes/JKFastNotificationProxy.swift
//
//  JKFastNotificationProxy.swift
//  JKNoticationHelper_Swift
//
//  Created by JackLee on 2021/9/14.
//

import Foundation
public class JKFastNotificationProxy {

    typealias JKFastNotificationBlock = (Notification) -> Void
    
    var notificationName: String
    var block: JKFastNotificationBlock
    
    init(name: String, block: @escaping JKFastNotificationBlock) {
        self.notificationName = name
        self.block = block
        NotificationCenter.default.addObserver(self, selector: #selector(receivedNotifcation(_:)), name: NSNotification.Name.init(notificationName), object: nil)
    }
    
    @objc func receivedNotifcation(_ noti: Notification){
        self.block(noti)
    }
    
    deinit {
        NotificationCenter.default.removeObserver(self)
    }
}
